### DISCUSSION TITLE
Route not changing

### DISCUSSION THREAD

**Discussion by sebastian-alupoaei** [eclipse/mosaic#385] [Q&A] [answered]
[Centru_Tudor.zip](https://github.com/eclipse/mosaic/files/14814331/Centru_Tudor.zip)
[NotificareAccident2.zip](https://github.com/eclipse/mosaic/files/14814332/NotificareAccident2.zip)
Hello, I tried to make a scenario similar to Barnim, but the cars do not change their route. Messages are sent, commands are received, but they remain on the same route. I can't figure out what could be wrong. Uploaded the whole scenario and Java class

**Comment by schwepmo** [comment]
Hey @sebastian-alupoaei,
sorry for the delayed answer.
Without running the scenario on my own machine I would suggest that you try to debug and look at your `NotificareAccident2` class at line 322 and check if A) vehicles react at all to the send out DENMs and B) the NavigationModule is able to actually find a new route for the vehicles.

Feel free to add to this discussion if everything still seems fine.

**Reply by schwepmo** [reply]
Oh another potential cause, I just remembered, is that there isn't enough time for your vehicles to perform the route change. Meaning that even if an alternative route is found, it is to late for the vehicle to change it once it is propagated to SUMO. If this is what happens, you should find indications of it in the `Traffic.log`.

**Reply by sebastian-alupoaei** [reply]
Hi @schwepmo,
Firstly, thanks for your answer. Everything looks fine, (the same comparing with Barnim scenario's log) the cars could choose another route. This is how the log for veh_21 looks like:
```
2024-04-15 18:43:15,063 INFO  CentralNavigationComponent - Change to route 2 for vehicle veh_21.
2024-04-15 18:43:15,063 INFO  SumoGuiAmbassador - Receive VehicleRouteChange request at simulation time 174.012,192,715 s: vehicleId="veh_21", newRouteId=2, current edge: 22962250_247412044_9865773051
```
Moreover, I exaggerated the radius for sending messages to make sure that the cars receive the messages in time. Below I have shown an alternative route that the cars could have used.

![image](https://github.com/eclipse/mosaic/assets/79857889/2cf9cdbb-687e-4399-8261-7fa1c8b3bb15)

**Reply by schwepmo** [reply, marked as answer]
![grafik](https://github.com/eclipse/mosaic/assets/39306374/ebc77037-b6a7-427c-8546-e154fd733991)

I've found a couple of potential issues, but I'm not sure if any of these will fix the behaviour:
1. In your server application the `HAZARD_ROAD` doesn't see to exist. Also the locations in the `environment_config.json` and your server don't quite seem to match.
2. In the image I marked with the red circle where you set your hazard area and with the green arrow I marked where you want your vehicles to go. In order for your vehicles to choose the marked route the cost function of the alternative route needs to be smaller than that of the previous route even with the increased cost. As the affected road seems to be quite short and the intended detour seems to be quite a bit longer I can imagine, that even with the hazard zone the previous route is faster.

If you are running MOSAIC from within your IDE (which I would highly recommend) I would check if the generated route "2" follows your desired trajectory, i.e. by setting a breakpoint in the `CentralNavigationComponent` class at line 329 and check the edge ids.
I have the hunch that this might not be the issue and that you need to further "worsen" the condition on the affected road, either by reducing the speed or choosing a longer segment.

**Reply by sebastian-alupoaei** [reply]
It turned out that reducing the speed was the only solution. Thank you very much.

**Reply by schwepmo** [reply]
I'm glad you were able to figure out a solution. Good luck on your future MOSAIC adventures, feel free to reach out whenever you encounter any further problems!
